### ISSUE TITLE
Can addHeader handle entries other than content-type?

### ISSUE THREAD

**Issue by aferust** [trikko/serverino#14]
I need CORS stuff. it seems `output.addHeader("Access-Control-Allow-Origin", "*");` have no effect when I try to use fetch func of react? I took a look at the sources and if I am not wrong it just receive and sends the header stuff without touching it (_internal._headers ~= KeyValue(key.toLower, value);). But in my case somehow those don't have any effect.

```d
output.addHeader("content-type", "application/json");
output.addHeader("Access-Control-Allow-Origin", "*");
output.addHeader("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS");
output.addHeader("Access-Control-Allow-Headers", "Origin, X-Requested-With, content-type, Accept, Authorization");

output ~= jsonOut.toString;
```

from chrome debug console:
`Access to fetch at 'http://localhost:8080/newuser' from origin 'http://localhost:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.`

**Comment by trikko** [comment]
I’m not sure I understood: have you tried to see what headers Serverino responds with? Does Serverino send all the headers that you had added?

Serverino adds every header you send, even custom ones, like "X-my-custom-header".

**Comment by aferust** [comment]
Dear Andrea, I see. For now, I just use no-cors to quickly complete my little project. I will revisit this issue later and will inform you about it. Thank you.

**Comment by aferust** [comment]
Dear Andrea, my cors problem was resolved by adding "Content-Type": "multipart/form-data" to the fetch function's header. dunno why it is related to Cors barrier.  it's been a long time to cope with web programming. 

This means there is not an issue related to serverino, since output.addHeader("Access-Control-Allow-Origin", "*") does the expected thing.

**Comment by aferust** [comment]
Hello Andrea, I want to ask some questions (I am not in a rush, take your time please). Should Serverino have a handle mechanism for preflight requests? Or is it better to leave this kind of responsibility to the user? However, this is a small tool, and I prefer to keep this position. Maybe it has already one that I failed to figure out. I didn't even know that the preflight request thing existed. I had problems with fetch for this reason. I had to add some code to the beginning of every endpoint func body. mixin(Preflight!());

```d
template Preflight()
{
    const char[] Preflight = "
    if (request.method == typeof(request.method).Options) {
        // Respond with HTTP OK status for preflight requests
        output.status = 200;
        return;
    }";
}
```

My typical endpoint is like this. I wish I could access the Output parameter in UDAs to remove the auth check from the function body.
```d
...
@endpoint
@route!(r => r.uri=="/loggedin/setuserpass")
void changePassword(Request request, Output output){
	setOutHeaders(output);
	mixin(Preflight!());

	JSONValue res;
	if(performBasicAuth(request, output)){
		immutable json = handleInJson(request.body.data);
		string newpassword = json["password"].str;

		auto response = responseWithNewToken();
		sessionUser.hashPassword(newpassword);
		updatePasswordDB(sessionUser); // sessionUser is just a global class obj. am I doing right? do we need some thing like g in flask : https://flask.palletsprojects.com/en/2.3.x/appcontext/
		res = JSONValue( ["result": "1", "token": response["token"].toString]);
	}else{
		res = JSONValue( ["result": "0", "err msg": "wrong user info"]);
	}
	output ~= res;	
}
...
bool performBasicAuth(ref Request request, ref Output output) {
    string authHeader = request.header.read("authorization");

    if (authHeader.empty || !authHeader.startsWith("Basic ")) {
        output.status = 401;
        return false;
    }
    
    string base64Credentials = authHeader.split(" ")[1];
    auto credentials = Base64.decode(base64Credentials).split(":");
    
    string username = cast(string)credentials[0];
    string password = cast(string)credentials[1];

    if (verifyPassword(username, password)) {
        output.status = 200;
        return true;
    } else {
        output.status = 401;
        return false;
    }
}
...
void setOutHeaders(ref Output output){
	output.addHeader("Access-Control-Allow-Origin", "*");
	output.addHeader("Access-Control-Allow-Credentials", "true");
	output.addHeader("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS");
	output.addHeader("Access-Control-Allow-Headers", "Authorization, Content-Type");
	output.addHeader("Content-Type", "application/json");
}
```
In the university department I work for, we have a windows server 2008 (damn old one right?). I want to setup things and run my serverino API on that computer. nginx is not an option on windows IMHO. What do you recommend for it?

Thanks in advance 👏

**Comment by aferust** [comment]
Thank you. I forgot priority feature of the library. probably I will refactor my code base accordingly.

**Comment by aferust** [comment]
Hi Andrea,

It is me, again 😃. I am writing a program like Serve of Martin Nowak but
using serverino. My code draft runs well and can successfully serve a react
source folder. I am going to publish it if I can resolve some issues. I
would make a pull request to Serverino repo if you like the idea. Here is
my problem. I believe that ServerinoConfig setup() runs multiple times (for
each worker?). I expect it to run at the exe startup once. For the first
run, args are passed, then the second one, and others run without args.
This situation exists even with Runtime.args(). this prevents me from
nicely handling program arguments like ".\servefolder path port". Am
I missing something?

**Comment by trikko** [comment]
Probably are you missing "@onWorkerInit"?

**Comment by aferust** [comment]
it is obvious that onServerInit is called multiple times, and only the
first one has args:

setup() called with args:
["D:\\projects\\d_projects\\servefolder\\servefolder.exe",
"D:\\projects\\deneyler-gui\\build", "8080"]
- [018836] [i] 2024-02-29 22:21 [daemon.d:0210] Daemon started.
- [018836] [i] 2024-02-29 22:21 [daemon.d:0233] Listening on
http://0.0.0.0:8080/
setup() called with args:
["D:\\projects\\d_projects\\servefolder\\servefolder.exe"]
D [007736] [l] 2024-02-29 22:21 [worker.d:0084] Worker started.
setup() called with args:
["D:\\projects\\d_projects\\servefolder\\servefolder.exe"]
D [010736] [l] 2024-02-29 22:21 [worker.d:0084] Worker started.
setup() called with args:
["D:\\projects\\d_projects\\servefolder\\servefolder.exe"]
D [007508] [l] 2024-02-29 22:21 [worker.d:0084] Worker started.
setup() called with args:
["D:\\projects\\d_projects\\servefolder\\servefolder.exe"]
D [004772] [l] 2024-02-29 22:21 [worker.d:0084] Worker started.

**Comment by trikko** [comment]
Yes it is called multiple times, but you should not put initialization code inside, you should use it only to return a config.

Init code should be put inside ondaemonstart (main process start) or onworkerstart (when a worker is started or restarted)

Probably the best way is to use runtime.args inside ondaemonstart (that runs before any worker)

**Comment by trikko** [comment]
If you need to pass something between main process (Daemon) and worker try to use envvar, it should work

**Comment by trikko** [comment]
Eg:
```
environment["foo"] = "bar";
```

**Comment by trikko** [comment]
Check this:

```d
module app;

import serverino;
import std;

mixin ServerinoMain;

@onServerInit ServerinoConfig configure()
{
	return ServerinoConfig
		.create()
		.addListener("0.0.0.0", 8080)
		.setWorkers(4);
}

@onDaemonStart void start()
{
	environment["path"] = "/tmp";
	environment["data"] = `{"name": "John", "age": 30}`;
}


void dump(Request request, Output output)
{
	output.addHeader("content-type", "text/plain");
	output ~= "Path: " ~ environment["path"] ~ "\n";
	output ~= "Data: " ~ environment["data"] ~ "\n";
}
```

**Comment by aferust** [comment]
Hello Andrea,

Thanks for the tip. I restructured the code accordingly. Do you think it is
ready to put on the dub registry? Is there something that I missed and not
covered for a wide range of use cases? Do you want its ownership? If so, I
can make it a subproject in Serverino with PR.

https://github.com/aferust/servefolder

**Comment by trikko** [comment]
Nice!

Some features I would add:
- Do not serve (and list?) hidden files by default (f.e. files starting with ".")
- Do not serve (and list?) files serverino can't read (without read permission)
- Do not serve (and follow?) symlinks (a folder can exit from your root!)
- Why not using --port --serve-hidden-files and other options like this? (see std.getopt)
- `your-ip/index.html` hides a index.html file in the root. Maybe you could serve files unsing  `your-ip/?file=/index.html` ?

**Comment by aferust** [comment]
> Do not serve (and list?) hidden files by default (f.e. files starting
with ".")
this one is already covered. see
https://github.com/aferust/servefolder/blob/main/source/dirfileops.d#L27

> Why not using --port --serve-hidden-files and other options like this?
(see std.getopt)
I was using std.getopt at first. then i hit the args issue and just used
manual handling for args.now it is resolved and i can easily reimplement it.

> Do not serve (and list?) files serverino can't read (without read
permission)
> Do not serve (and follow?) symlinks (a folder can exit from your root!)
i will cover these.

> your-ip/index.html hides a index.html file in the root. Maybe you could
serve files unsing your-ip/?file=/index.html ?
i will consider that one too

**Comment by trikko** [comment]
First one: you do not list these files, but I can still access them writing the right url, right?

**Comment by aferust** [comment]
ahh yes. i shoulda use folder.baseName.startsWith("."). Probably i will
need a blackList function to handle all similar cases in one united place

**Comment by trikko** [comment]
Or maybe split the whole requested path and check if there's a part starting with .
/this/request/.should/not/pass

**Comment by trikko** [comment]
Two points:
- It doesn't work if an index doesn't exists? Usually they create an index page with a (clickable) list on link to internal files like https://ftp.debian.org/debian/tools/ (--disable-listing)
- A better way to serve a file is using output.serveFile()  see: https://trikko.github.io/serverino/serverino/interfaces/Output.serveFile.html

**Comment by aferust** [comment]
I always find this vulnerable https://ftp.debian.org/debian/tools/, listing
a folder browser. Maybe better to use an option for it.

i am using output.serveFile() in my little project I mentioned before. Yes,
in the next release, we can serve those with serveFile. Can you help me
prepare a static immutable array containing extensions suitable to be
downloadable or to serve with serveFile? or just tell me the types. I can
cover your points in the next release.

**Comment by trikko** [comment]
You can serve anything, also html, what's the problem? It should be even more efficient than output ~= readText().

output ~= something is useful for dynamic webpages, not for serving static pages

**Comment by aferust** [comment]
hey Andera, you were right. I implemented a little folder browser for
default index, allowing me to reveal many bugs. for instance, /path/folder/
and /path/folder/ must be handled differently. I am working on it. I will
push when it is complete.

Thanks

**Comment by aferust** [comment]
*: /folder/antherfolder/ and  /folder/antherfolder

**Comment by trikko** [comment]
I don't get a thing: why do you create a full list of files on start?

Why don't you simply pass the base folder and check if requested file exist when required?

**Comment by aferust** [comment]
i thought parsing the entire folder once at the beginning of the server
init (ondaemonstart) was more efficient than doing one path checks for each
request.  But i am reevaluating this because I am learning how things work.

2 Mar 2024 Cmt 12:51 tarihinde Andrea Fontana ***@***.***>
şunu yazdı:

**Comment by aferust** [comment]
imagine something crazy. Load everything to the  heap at the beginning all
at once.  you can after even delete the folder you serve. this requires a
huge amount of ram, but you serve like light speed, right?, no disk reads
for the resources you serve, just thinking

2 Mar 2024 Cmt 13:05 tarihinde Ferhat Kurtulmuş ***@***.***> şunu
yazdı:

**Comment by trikko** [comment]
Maybe, but it takes a lot of time during startup and you have to restart
server if you change a file.

And os has also a disk cache.

And anyway connection speed is a bottleneck.


Il sab 2 mar 2024, 11:30 aferust ***@***.***> ha scritto:

**Comment by aferust** [comment]
hello Andrea,

I restructured the code base. These changes make me feel more comfortable
now. it now includes a simple directory browser (displays if no index
file). I tried output.serveFile but it immediately fires a download on the
browser. We should only handle predefined extensions with serveFile.

**Comment by trikko** [comment]
Are you sure it fires a download? It should not. Are you adding some extra headers? Do you have an example?

**Comment by aferust** [comment]
it fires a download if there is no content type attached, so I will
use serveFile, ok. See, I added something like .htpasswd of Apache.
